src/axom/quest/MarchingCubes.hpp

```diff
@@ -0,0 +1,295 @@
+    coordinates of the intersection.
+  */
+  void linear_interp(int edgeIdx,
+                     const double *xx,
+                     const double *yy,
+                     const double *zz,
+                     const double *nodeValues,
+                     double *xyz);
+
+  /*!
+    @brief Compute the index to look up the surface topology in a cell.
+    @param f Scalar function values at the 4 or 8 nodes of the cell.
+  */
+  int computeIndex(const double *f);
+
+};  // class MarchingCubesSingleDomain
+
+/*!
+ * \@brief Class implementing marching cubes algorithm.
+ *
+ * \sa MarchingCubesSingleDomain
+ */
+class MarchingCubes
+{
+public:
+  /*!
+   * \brief Constructor sets up computational mesh and data for running the
+   * marching cubes algorithm.
+   *
+   * \param [in] bpMesh Blueprint multi-domain mesh containing scalar field.
+   * \param [in] coordsetName Name of blueprint point coordinate set.
+   * \param [in] maskField Cell-based std::int32_t mask field.  If provided,
+   *             cells where this field evaluates to false are skipped.
+   *
+   * Some data from \a bpMesh may be cached by the constructor.
+   * Any change to it after the constructor leads to undefined behavior.
+   *
+   * The mesh coordinates should be contiguous.  See
+   * conduit::blueprint::is_contiguous().  In the future, this
+   * requirement may be relaxed, possibly at the cost of a
+   * transformation and storage of the temporary contiguous layout.
+   */
+  MarchingCubes(const conduit::Node &bpMesh,
+                const std::string &coordsetName,
+                const std::string &maskField = {});
+
+  /*!
+    @brief Set the field containing the nodal function.
+    \param [in] fcnField Name of node-based scalar function values.
+  */
+  void set_function_field(const std::string &fcnField);
+
+  /*!
+    @brief Set the output surface mesh object.
+  */
+  void set_output_mesh(axom::mint::Mesh *surfaceMesh);
+
+  /*!
+    @brief Save the originating 1D cell index as an int in the specified
+    cell-centered field of the mesh.
+
+    If the field doesn't exist, it will be created.  To disable, set
+    \a domainIdField to empty.
+  */
+  void set_cell_id_field(const std::string &cellIdField)
+  {
+    m_cellIdField = cellIdField;
+    for(auto &s : m_sd)
+    {
+      s->set_cell_id_field(cellIdField);
+    }
+  }
+
+  /*!
+    @brief Save the originating domain index as an int in the specified
+    cell-centered field of the mesh.
+
+    If the field doesn't exist, it will be created.  To disable, set
+    \a domainIdField to empty.
+  */
+  void set_domain_id_field(const std::string &domainIdField)
+  {
+    m_domainIdField = domainIdField;
+  }
+
+  /*!
+ * \brief Computes the iso-surface.
+ *
+ * \param [in] contourVal iso-contour value
+ */
+  void compute_iso_surface(double contourVal = 0.0);
+
+private:
+  //! @brief Single-domain implementations.
+  axom::Array<std::shared_ptr<MarchingCubesSingleDomain>> m_sd;
```

Comment:
Is it possible for us to make this `std::unique_ptr<MarchingCubesSingleDomain>`? We should only need `shared_ptr` for cases where ownership isn't actually clear.